README
NAME
    GIS::Distance - Calculate geographic distances.

SYNOPSIS
      use GIS::Distance;
  
      my $gis = GIS::Distance->new();
  
      $gis->formula( 'Polar' );  # Optional, default is Haversine.
  
      my $distance = $gis->distance( $lat1,$lon1 => $lat2,$lon2 );
  
      print $distance->meters();

DESCRIPTION
    This perl library aims to provide as many tools to make it as simple as
    possible to calculate distances between geographic points, and anything
    that can be derived from that.

METHODS
  distance
      my $distance = $gis->distance( $lat1,$lon1 => $lat2,$lon2 );

    Returns a Class::Measure::Length object for the distance between the two
    degree lats/lons. The distance is calculated using whatever formula the
    object is set to use.

ATTRIBUTES
  formula
    This is an object who's class inherits from GIS::Distance::Formula. This
    object is used to calculate distance. The formula may be specified as
    either a blessed object, or as a string, such as "Haversine" or any of
    the other formulas.

    If you specify the formula as a string then a few different class names
    will be searched for. So, if you did:

      $gis->formula( 'Haversine' );

    Then this list of packages would automatically be looked for. The first
    one that exists will be created and used:

      GIS::Distance::Formula::Haversine::Fast
      GIS::Distance::Formula::Haversine
      Haversine

    If you are using your own custom formula class make sure it extends()
    (Moose) the GIS::Distance::Formula class.

    Note that a ::Fast version of the class will be looked for first. By
    default the ::Fast versions of the formulas, written in C, are not
    available and the pure perl ones will be used instead. If you would like
    the ::Fast formulas then install GIS::Distance::Fast and they will be
    automatically used.

SEE ALSO
    GIS::Distance::Fast - C implmentation of some of the formulas shipped
    with GIS::Distance. This greatly increases the speed at which distance
    calculations can be made.

FORMULAS
    GIS::Distance::Formula::Cosine

    GIS::Distance::Formula::GeoEllipsoid

    GIS::Distance::Formula::GreatCircle

    GIS::Distance::Formula::Haversine

    GIS::Distance::Formula::MathTrig

    GIS::Distance::Formula::Polar

    GIS::Distance::Formula::Vincenty

TODO
    *   Create a GIS::Coord class that represents a geographic coordinate.
        Then modify this module to accept input as either lat/lon pairs, or
        as GIS::Coord objects.

    *   Create an extension to DBIx::Class with the same goal as
        Geo::Distance's closest() method.

    *   Write a super accurate formula module called GIS::Distance::Geoid.
        Some very useful info is at <http://en.wikipedia.org/wiki/Geoid>.

BUGS
    Both the GIS::Distance::Formula::GreatCircle and
    GIS::Distance::Formula::Polar formulas are broken. Read their respective
    man pages for details.

TEST COVERAGE
      ---------------------------- ------ ------ ------ ------ ------ ------ ------
      File                           stmt   bran   cond    sub    pod   time  total
      ---------------------------- ------ ------ ------ ------ ------ ------ ------
      blib/lib/GIS/Distance.pm      100.0    n/a    n/a  100.0  100.0   24.0  100.0
      ...b/GIS/Distance/Formula.pm   75.0    n/a    n/a   66.7  100.0    1.9   75.0
      ...istance/Formula/Cosine.pm  100.0    n/a    n/a  100.0  100.0    5.9  100.0
      ...e/Formula/GeoEllipsoid.pm  100.0    n/a    n/a  100.0  100.0    3.6  100.0
      ...ce/Formula/GreatCircle.pm  100.0    n/a    n/a  100.0  100.0    5.9  100.0
      ...ance/Formula/Haversine.pm  100.0    n/a    n/a  100.0  100.0    9.1  100.0
      ...tance/Formula/MathTrig.pm  100.0    n/a    n/a  100.0  100.0    2.4  100.0
      ...Distance/Formula/Polar.pm  100.0    n/a    n/a  100.0  100.0    2.7  100.0
      ...tance/Formula/Vincenty.pm  100.0   50.0   50.0  100.0  100.0   44.6   93.1
      Total                          98.8   50.0   50.0   97.2  100.0  100.0   96.7
      ---------------------------- ------ ------ ------ ------ ------ ------ ------

AUTHOR
    Aran Clary Deltac <bluefeet@cpan.org>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension GIS-Distance.

{{$NEXT}}

0.20 2023-06-07T19:24:59Z

 - Minor metadata and boilerplate updates.
 - Fix typo in POD.

0.19 2021-01-30T06:07:58Z

 - Changed licensing terms to be the same as Perl.
 - Provide a .editorconfig instead of .lvimrc.

0.18 2019-05-10T20:31:31Z

 - Switch to the GNU General Public License version 3.
 - Fixed pod error as reported by CPANTS.
 - Documentation edits.

0.17 2019-03-17T17:01:31Z

 - Finalize the new internal formula interface.

0.16 2019-03-16T22:19:33Z

 - Renamed the args and module attributes to formula_args
   and formula_module to make them less generic and a bit
   more accurate.
 - Fix a pretty bad math typo in the SPEED section.
 - Lots of other documentation edits.

0.15 2019-03-13T06:06:09Z

 - Support using Geo::Point as arguments to distance().
 - Lots of documentation edits.
 - Moved TODO section into GitHub issues.
 - Added a benchmark to the SPEED section.
 - Made a bunch of improvements to the author tools.

0.14 2019-03-10T05:04:54Z

 - Add GIS::Distance::ALT formula.
 - Removed distance_km().

0.13 2019-03-09T12:32:55Z

 - Add abs() to Haversine.
 - Added GIS::Distance::Null, the fastest formula yet.

0.12 2019-03-08T18:32:13Z

 - Added the distance_metal() method to GIS::Distance.
 - Various documentation edits, including a new SPEED section.

0.11 2019-03-07T22:23:02Z

 - Support the GIS_DISTANCE_PP environment variable.
 - Don't support older ::Formula modules, makes no sense and
   they wouldn't work anyways.
 - Declare Carp dep.
 - Lots and lots of documentation edits.
 - Recommend the newer GIS::Distance::Fast.

0.10 2019-03-07T16:28:48Z

 - WARNING: The GIS::Distance object is now immutable, thus the
   formula can no longer be set with the formula attribute!
 - Moved GIS::Distance::GeoEllipsoid to a separate distro.
 - Added the distance_km() method to GIS::Distance.
 - Removed Moo and Type::Tiny, all unecessary, simple OO.
 - Move GIS::Distance::Formula:: modules to GIS::Distance::.
 - Migrate build tooling from Dist::Zilla to Minilla.

0.09 2015-06-11

 - Move away from Any::Moose to Moo (yay!).
 - Better formula loading logic.
 - Support single-arg (formula) GIS::Distance instantiation.

0.08 2012-03-23

 - Release with Dist::Zilla.
 - Fix Great Circle formula to use ** instead of ^.

0.07 2010-02-02

 - Use Any::Moose instead of Moose directly.
 - Declare namespace::autoclean dependency.

0.06 2010-01-30

 - Minor build updates to include some extra info (github, etc).

0.05 2010-01-30

 - Speed improvements under Moose (now uses immutable).

0.04 2009-06-29

 - Fixed for latest Moose.
 - Fixed for latest Class::Measure.

0.03

 - Fixed for latest Moose.
 - Reduced the README to a one-liner.
 - Added docs to GIS::Distance::Formula.
 - Fixed some documentation typos that were using GID instead of GIS.
 - Fixed a typo that mispelled "formula" as "formuka".
 - Refer to GIS::Distance::Fast in the SEE ALSO section.
 - Added a one liner to the SYNOPSIS showing how to used the returned distance object.
 - Added a TEST COVERAGE section with output from Devel::Cover.

0.02 2008-03-16

 - Added Geoid to the TODO section.
 - Using Module::Install now.
 - Moved all formulas in to the GIS::Distance::Formula namespace.
 - Using Moose for all OO now.
 - Added tests (bout time!).
 - Added support for the up-and-coming ::Fast:: modules.
 - Added (BROKEN) to the abstract for the GreatCircle and Polar formulas.
 - Fixed POD testing.
 - Changed version scheme to use the simple x.xx format.

0.01001 2006-09-20

 - Added basic META.yml.
 - Geo::Ellipsoid support.
 - Added a TODO section.
 - Various bug fixes to the Vincenty formula.
 - GreatCircle formula marked as broken.
 - Added a dev script for graphing the deviations in the formulas.

0.01000 2006-09-19

 - Renamed from Geo::Distance to GIS::Distance.
 - Moved distance calculations in to their own modules.
 - Use Class::Measure::Length to handle distance return values.
 - Test POD syntax.
 - Forked from Geo::Distance 0.11.


lib/GIS/Distance.pm
package GIS::Distance;
use 5.008001;
use strictures 2;
our $VERSION = '0.19';

sub new {
    my ($class, $formula, @args) = @_;

    $formula ||= 'Haversine';

    my @modules;
    push @modules, "GIS::Distance::Fast::${formula}"
        unless $ENV{GIS_DISTANCE_PP} or $ENV{GEO_DISTANCE_PP};
    push @modules, "GIS::Distance::$formula";
    push @modules, $formula;

    foreach my $module (@modules) {
        next if !_try_load_module( $module );
        next if !$module->isa('GIS::Distance::Formula');

        return $module->new( @args );
    }

    die "Cannot find a GIS::Distance::Formula class for $formula";
};

my %tried_modules;

sub _try_load_module {
    my ($module) = @_;

    return $tried_modules{ $module }
        if defined $tried_modules{ $module };

    $tried_modules{ $module } = 1;
    my $ok = eval( "require $module; 1" );
    return 1 if $ok;

    $tried_modules{ $module } = 0;
    die $@ if $@ !~ m{^Can't locate};
    return 0;
}

1;
__END__

=encoding utf8

=head1 NAME

GIS::Distance - Calculate geographic distances.

=head1 SYNOPSIS

    use GIS::Distance;
    
    # Use the GIS::Distance::Haversine formula by default.
    my $gis = GIS::Distance->new();
    
    # Or choose a different formula.
    my $gis = GIS::Distance->new( 'Polar' );
    
    # Returns a Class::Measure object.
    my $distance = $gis->distance( $lat1, $lon1, $lat2, $lon2 );
    
    print $distance->meters();

=head1 DESCRIPTION

This module calculates distances between geographic points on, at the moment,
planet Earth.  Various L</FORMULAS> are available that provide different levels
of accuracy versus speed.

L<GIS::Distance::Fast>, a separate distribution, ships with C implementations of
some of the formulas shipped with GIS::Distance.  If you're looking for speed
then install it and the ::Fast formulas will be automatically used by this module.

=head1 METHODS

=head2 distance

    my $distance = $gis->distance( $lat1, $lon1, $lat2, $lon2 );
    
    my $point1 = Geo::Point->latlong( $lat1, $lon1 );
    my $point2 = Geo::Point->latlong( $lat2, $lon2 );
    my $distance = $gis->distance( $point1, $point2 );

Takes either two decimal latitude and longitude pairs, or two L<Geo::Point>
objects.

Returns a L<Class::Measure::Length> object for the distance between the
two degree lats/lons.

See L</distance_metal> for a faster, but less feature rich, method.

=head2 distance_metal

This works just like L</distance> except for:

=over

=item *

Does not accept L<Geo::Point> objects.  Only decimal latitude and longitude
pairs.

=item *

Does not return a L<Class::Measure> object.  Instead kilometers are always
returned.

=item *

Does no argument checking.

=item *

Does not support formula arguments which are supported by at least the
L<GIS::Distance::GeoEllipsoid> formula.

=back

Calling this gets you pretty close to the fastest bare metal speed you can get.
The speed improvements of calling this is noticeable over hundreds of thousands of
iterations only and you've got to decide if its worth the safety and features
you are dropping.  Read more in the L</SPEED> section.

=head1 ARGUMENTS

    my $gis = GIS::Distance->new( $formula );

When you call C<new()> you may pass a partial or full formula class name as the
first argument.  The default is C<Haversive>.

If you pass a partial name, as in:

    my $gis = GIS::Distance->new( 'Haversine' );

Then the following modules will be looked for in order:

    GIS::Distance::Fast::Haversine
    GIS::Distance::Haversine
    Haversine

Install L<GIS::Distance::Fast> to get access to the C<Fast::> (XS) implementations
of the formula classes.

You may globally disable the automatic use of the C<Fast::> formulas by setting
the C<GIS_DISTANCE_PP> environment variable.  Although, its likely simpler to
just provide a full class name to get the same effect:

    my $gis = GIS::Distance->new( 'GIS::Distance::Haversine' );

=head1 SPEED

Not that this module is slow, but if you're doing millions of distance
calculations a second you may find that adjusting your code a bit may
make it faster.  Here are some options.

Install L<GIS::Distance::Fast> to get the XS variants for most of the
PP formulas.

Use L</distance_metal> instead of L</distance>.

Call the undocumented C<_distance()> function that each formula class
has.  For example you could bypass this module entirely and just do:

    use GIS::Distance::Fast::Haversine;
    my $km = GIS::Distance::Fast::Haversine::_distance( @coords );

The above would be the ultimate speed demon (as shown in benchmarking)
but throws away some flexibility and adds some foot-gun support.

Here's a benchmarks for these options:

    2019-03-13T09:34:00Z
    GIS::Distance 0.15
    GIS::Distance::Fast 0.12
    GIS::Distance::Fast::Haversine 0.12
    GIS::Distance::Haversine 0.15
                                                                 Rate
    PP Haversine - GIS::Distance->distance                   123213/s
    XS Haversine - GIS::Distance->distance                   196232/s
    PP Haversine - GIS::Distance->distance_metal             356379/s
    PP Haversine - GIS::Distance::Haversine::_distance       385208/s
    XS Haversine - GIS::Distance->distance_metal             3205128/s
    XS Haversine - GIS::Distance::Fast::Haversine::_distance 8620690/s

You can run your own benchmarks using the included C<author/bench>
script.  The above results were produced with:

    author/bench -f Haversine

The slowest result was about C<125000/s>, or about C<8ms> each call.
This could be a substantial burden in some contexts, such as live HTTP
responses to human users and running large batch jobs, to name just two.

In conclusion, if you can justify the speed gain, switching to
L</distance_metal> and installing L<GIS::Distance::Fast> looks to be an
ideal setup.

As always with performance and benchmarking, YMMV.

=head1 COORDINATES

When passing latitudinal and longitudinal coordinates to L</distance>
they must always be in decimal degree format.  Here is some sample code
for converting from other formats to decimal:

    # DMS to Decimal
    my $decimal = $degrees + ($minutes/60) + ($seconds/3600);
    
    # Precision Six Integer to Decimal
    my $decimal = $integer * .000001;

If you want to convert from decimal radians to degrees you can use L<Math::Trig>'s
rad2deg function.

=head1 FORMULAS

These formulas come bundled with this distribution:

=over

=item *

L<GIS::Distance::ALT>

=item *

L<GIS::Distance::Cosine>

=item *

L<GIS::Distance::GreatCircle>

=item *

L<GIS::Distance::Haversine>

=item *

L<GIS::Distance::MathTrig>

=item *

L<GIS::Distance::Null>

=item *

L<GIS::Distance::Polar>

=item *

L<GIS::Distance::Vincenty>

=back

These formulas are available on CPAN:

=over

=item *

L<GIS::Distance::Fast/FORMULAS>

=item *

L<GIS::Distance::GeoEllipsoid>

=back

=head1 AUTHORING

Take a look at L<GIS::Distance::Formula> for instructions on authoring
new formula classes.

=head1 SEE ALSO

=over

=item *

L<Geo::Distance> - Is deprecated in favor of using this module.

=item *

L<Geo::Distance::Google> - While in the Geo::Distance namespace, this isn't
actually related to Geo::Distance at all.  Might be useful though.

=item *

L<GIS::Distance::Lite> - An old fork of this module, not recommended.

=item *

L<Geo::Distance::XS> - Used to be used by L<Geo::Distance> but no longer is.

=item *

L<Geo::Ellipsoid> - Or use L<GIS::Distance::GeoEllipsoid> for a uniform
interface.

=item *

L<Geo::Inverse> - Does some distance calculations, but seems less than useful
as all the code looks to be taken from L<Geo::Ellipsoid>.

=back

=head1 SUPPORT

Please submit bugs and feature requests to the
GIS-Distance GitHub issue tracker:

L<https://github.com/bluefeet/GIS-Distance/issues>

=head1 AUTHOR

    Aran Clary Deltac <bluefeet@gmail.com>

=head1 CONTRIBUTORS

    Mohammad S Anwar <mohammad.anwar@yahoo.com>
    Florian Schlichting <fsfs@debian.org>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

